packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -360,111 +400,340 @@ contract AmmPool is IBlockReceiver {
             require(update.feeBips == feeBips, "INVALID_TX_DATA");
-            require(update.tokenWeight == tokens[i].weight, "INVALID_TX_DATA");
+            require(update.tokenWeight == (start ? 0 : ctx.tokens[i].weight), "INVALID_TX_DATA");
             // Now approve this AMM update
             update.validUntil = 0xffffffff;
-            bytes32 txHash = AmmUpdateTransaction.hashTx(exchangeDomainSeparator, update);
+            bytes32 txHash = AmmUpdateTransaction.hashTx(ctx.exchangeDomainSeparator, update);
             exchange.approveTransaction(address(this), txHash);
-            numTransactionsConsumed++;
-            // AMM account balance now available onchain
-            ammBalancesInAccount[i] = update.balance;
-            ammBalances[i] = ammBalancesInAccount[i];
+            ctx.numTransactionsConsumed++;
+            if (start) {
+                // AMM account balance now available onchain
+                ctx.ammBalancesInAccount[i] = update.balance;
+                ctx.ammBalances[i] = ctx.ammBalancesInAccount[i];
+            }
         }
+    }
 
-        // Process work in the queue
-        for (uint n = queuePos; n < queuePos + auxData.numItems; n++) {
-            QueueItem memory item = queue[n];
-            // TODO: all the necessary pool logic, mint/burn LP tokens,...
-            if (item.workType == WorkType.JOIN) {
-                PoolJoin memory join = item.join;
-                for (uint i = 0; i < tokens.length; i++) {
-                    uint amount = join.maxAmountsIn[i];
-                    ammBalances[i] = ammBalances[i].add(amount);
-                    // Unlock the amount locked for this join
-                    locked[item.owner][tokens[i].addr] = locked[item.owner][tokens[i].addr].sub(join.maxAmountsIn[i]);
-                    // Make the amount unavailable for withdrawing
-                    balance[item.owner][tokens[i].addr] = balance[item.owner][tokens[i].addr].sub(amount);
-                }
-            } else if (item.workType == WorkType.EXIT) {
-                PoolExit memory exit = item.exit;
-                for (uint i = 0; i < tokens.length; i++) {
-                    uint amount = exit.minAmountsOut[i];
-                    ammBalances[i] = ammBalances[i].sub(amount);
-                    // Make the amount available for withdrawing
-                    balance[item.owner][tokens[i].addr] = balance[item.owner][tokens[i].addr].add(amount);
+    function processJoin(
+        Context  memory ctx,
+        PoolJoin memory join,
+        bytes    memory signature
+        )
+        internal
+    {
+        bytes32 poolTxHash = hashPoolJoin(ctx.DOMAIN_SEPARATOR, join);
+        if (signature.length == 0) {
+            require(queue[queuePos].txHash == poolTxHash, "NOT_APPROVED");
+            delete queue[queuePos];
+            queuePos++;
+        } else {
+            require(poolTxHash.verifySignature(join.owner, signature), "INVALID_SIGNATURE");
```

Comment:
Even if we authenticate the request using signature, should we also delete the queued request by doing `delete queue[queuePos]`

Reply:
Maybe we can delete the `queuePos` variable and always use `queue[0]` and `delete queue[0]`

Reply:
> Even if we authenticate the request using signature, should we also delete the queued request by doing delete queue[queuePos]

In most cases there won't be any queue item for the pool transaction. Just a signature. Like with the exchange, a pool transaction can either be authorized with a signature or with an onchain approval.

> Maybe we can delete the queuePos variable and always use queue[0] and delete queue[0]

I don't think `delete queue[0]` pushes the next queue item to `[0]`.

I don't really like the queue, it's there because joins from layer 1 need to be locked up but if they need to be locked we could also just deposit to the exchange directly (but this will cost more gas because things like batching the deposits won't be possible...).